### PR TITLE
Refactor type parsing grammar to allow tokens as field names

### PR DESCRIPTION
**Pull request by pedroerp**
Summary:
Refactoring type parsing grammar to simplify it, fix a few bugs (e.g
disallow field names in top level types - check added unit tests), plus adding
support for common tokens like "array", "map", etc to be defined as type field
names.

Differential Revision: D52341881

### CODE REVIEW

File: velox/type/parser/TypeParser.ll
Comment (pedroerp) [inline velox/type/parser/TypeParser.ll:39]: Looks like presto does not like the "struct" keyword.

**Review by mbasmanova — APPROVED** [review]
@pedroerp Thank you for a quick fix. Accepting, but I'm not fluent in this code.

**Comment by pedroerp** [comment]
> @pedroerp Thank you for a quick fix. Accepting, but I'm not fluent in this code.

thanks for the quick review. I'll wait for @majetideepak to take a pass before I land it, as he has more context on this.

**Review by majetideepak — APPROVED** [review]
@pedroerp thanks for the detailed documentation and refactor.
Allowing `named_type` to have an optional name simplified the rules.

File: velox/type/parser/TypeParser.ll
Comment (majetideepak) [reply velox/type/parser/TypeParser.ll:39]: I think I followed the ANTLR grammar. We can always add this back later.

**Comment by pedroerp** [comment]
Thank you guys for the quick review :)

File: velox/type/parser/TypeParser.ll
Comment (pedroerp) [reply velox/type/parser/TypeParser.ll:39]: At this least this does not seem to work for me:

```
presto:de> select cast(null as struct(bigint));
Query 20231220_202534_73200_nyxqh failed: line 1:8: Unknown type: struct(bigint)

presto:de> select cast(null as row(bigint));
 _col0                
-------
 NULL  
(1 row)

```

**Pull request closed without merging**
